pkg/providers/v2/instances.go

```diff
@@ -0,0 +1,335 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package v2 is an out-of-tree only implementation of the AWS cloud provider.
+// It is not compatible with v1 and should only be used on new clusters.
+package v2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	v1 "k8s.io/api/core/v1"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/klog"
+)
+
+// newInstances returns an implementation of cloudprovider.InstancesV2
+func newInstances(region string, creds *credentials.Credentials) (cloudprovider.InstancesV2, error) {
+	awsConfig := &aws.Config{
+		Region:      aws.String(region),
+		Credentials: creds,
+	}
+	awsConfig = awsConfig.WithCredentialsChainVerboseErrors(true)
+
+	sess, err := session.NewSession(awsConfig)
+	if err != nil {
+		return nil, err
+	}
+	ec2Service := ec2.New(sess)
+
+	return &instances{
+		ec2: ec2Service,
+	}, nil
+}
+
+// EC2 is an interface defining only the methods we call from the AWS EC2 SDK.
+type EC2 interface {
+	DescribeInstances(request *ec2.DescribeInstancesInput) (*ec2.DescribeInstancesOutput, error)
+}
+
+// instances is an implementation of cloudprovider.InstancesV2
+type instances struct {
+	ec2 EC2
+}
+
+// InstanceExists indicates whether a given node exists according to the cloud provider
+func (i *instances) InstanceExists(ctx context.Context, node *v1.Node) (bool, error) {
+	var err error
+	if node.Spec.ProviderID == "" {
+		_, err = i.getInstanceByPrivateDNSName(ctx, node.Name)
+		if err == cloudprovider.InstanceNotFound {
+			return false, nil
+		}
+
+		if err != nil {
+			return false, err
+		}
+	} else {
+		_, err = i.getInstanceByProviderID(ctx, node.Spec.ProviderID)
+		if err == cloudprovider.InstanceNotFound {
+			return false, nil
+		}
+
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+// InstanceShutdown returns true if the instance is shutdown according to the cloud provider.
+func (i *instances) InstanceShutdown(ctx context.Context, node *v1.Node) (bool, error) {
+	var err error
+	var ret bool
+	if node.Spec.ProviderID == "" {
+		ret, err = i.instanceShutdownByPrivateDNSName(ctx, node.Name)
+	} else {
+		ret, err = i.instanceShutdownByProviderID(ctx, node.Spec.ProviderID)
+	}
+
+	return ret, err
+}
+
+func (i *instances) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprovider.InstanceMetadata, error) {
+	var err error
+	var ec2Instance *ec2.Instance
+	if node.Spec.ProviderID == "" {
+		//  TODO: support node name policy other than private DNS names
```

Comment:
not sure if I have enough contexts here, can you share more inputs? :) @andrewsykim

Reply:
gotcha, thanks for the contexts!

Reply:
After looking into kops and EKS in some detail, I don't think we would want "arbitrary" names, or at least if we do, it would have to come with huge caveats around node security.

In particular, EKS and kops-controller match a instance ID attestation with the privateDNSName, via an STS based authentication webhook in EKS via aws-iam-authenticator, or via issuance of a certificate with the privateDNSname as the node name with presentation of the AWS instance identity document in kops-controller's case (CAPI will likely also implement the latter).

In theory, this means we can securely support instanceID or privateDNSName, but beyond that, it would have to come with big "you don't want to use this" warnings.

@nckturner @justinsb

Reply:
Also, why would we use a ConfigMap vs. say ComponentConfig for the controller?

Reply:
EKS currently has a lot of built-in assumptions on private DNS.  That being said, instance ID is attractive because its guaranteed to be unique.  I'm ok with making it possible for other names, but we will need to be able to restrict it for EKS to stuff available to DescribeInstances, like @randomvariable said.  I don't think a ConfigMap is a good idea though.  Its seems like that suggests we want to allow on-the-fly reconfiguration, which for EKS we definitely wouldn't want.  Guessing, but if we allowed customers to make changes here then we'd probably want it protected behind an EKS API, which would be easier to do if it was configuration passed in via flags/ComponentConfig file on disk, and required a restart.

Reply:
> Also, why would we use a ConfigMap vs. say ComponentConfig for the controller?

I'm not particularly tied to using ConfigMap here, but if we use a config file it should ideally be yaml/json and not INI. I'm not sure ComponentConfig is relevant here since this is a config file read by aws-cloud-controller-manager, not a config for any of its options/flags. 

> EKS currently has a lot of built-in assumptions on private DNS. That being said, instance ID is attractive because its guaranteed to be unique. I'm ok with making it possible for other names, but we will need to be able to restrict it for EKS to stuff available to DescribeInstances

Sounds like there needs to be a broader discussion on this topic for sure. @nicolehanjing since we know for sure that we still want to support private DNS, let's get this PR only working with that and make sure we have a follow-up PR to support instance ID and other naming policies.

Reply:
gotcha! thanks for the info!